enterprise_access/apps/content_assignments/api.py

```diff
@@ -383,6 +373,118 @@ def _deduplicate_learner_emails_to_allocate(learner_emails):
+        # There's no case-insensitive IN query in Django, so we have to build up a big
+        # OR type of query.
+        email_lms_user_id = User.objects.filter(
+            _inexact_email_filter(email_chunk, field_name='email'),
+            lms_user_id__isnull=False,
+        ).values_list('email', 'lms_user_id')
+
+        # dict() on a list of 2-tuples treats the first elements as keys and second elements as values.
+        lms_user_ids_by_email.update(dict(email_lms_user_id))
+    return lms_user_ids_by_email
+
+
+def _get_existing_assignments_for_allocation(
+        assignment_configuration, learner_emails_to_allocate, content_key, lms_user_ids_by_email,
+):
+    """
+    Finds any existing assignments records related to the provided ``assignment_cofiguration``,
+    ``content_key``, and the learners the client has requested to allocate.
+    Learners are identified either via the `learner_email` field, or via the `lms_user_id` field
+    based on the provided ``learner_emails_to_allocate`` and the provided mapping
+    ``lms_user_ids_by_email``.
+    """
+    # Compose a set of all existing assignments related to the requested allocation.
+    # A set is a fine data structure because Django models are hashable on their PK.
+    existing_assignments = set()
+
+    # Fetch any existing assignments for all pairs of (learner email, content) in this assignment config.
+    assignments_for_emails_queryset = get_assignments_for_admin(
+        assignment_configuration, learner_emails_to_allocate, content_key,
+    )
+    existing_assignments.update(assignments_for_emails_queryset)
+
+    # Fetch existing assignments for all pairs of (known lms_user_id, content) in this assignment config.
+    assignments_for_lms_user_ids_queryset = get_assignments_for_configuration(
+        assignment_configuration,
+        lms_user_id__in=lms_user_ids_by_email.values(),
+        content_key=content_key,
+    )
+    existing_assignments.update(assignments_for_lms_user_ids_queryset)
+
+    return existing_assignments
+
+
+def _reallocate_assignment(
+    assignment, emails_by_lms_user_id, lms_user_ids_by_email, content_quantity,
+    allocation_batch_id, preferred_course_run_key,
+):
+    """
+    Modifies a ``LearnerContentAssignment`` record during the allocation flow.  The record
+    is **not** saved.
+    """
+    if assignment.lms_user_id:
+        # If the existing assignment already has an lms_user_id,
+        # it *might* have a retired/expired email address which we want to change
+        # based on our lookup of lms_user_id -> email.
+        assignment.learner_email = emails_by_lms_user_id.get(assignment.lms_user_id)
+    else:
+        # We still might not have an lms_user_id for this email
+        # address yet, but in that case, setting the lms_user_id
+        # as below is a no-op, which is fine.
+        assignment.lms_user_id = lms_user_ids_by_email.get(assignment.learner_email.lower())
```

Comment:
Everything except this part is a clean refactor/extraction of a helper function.